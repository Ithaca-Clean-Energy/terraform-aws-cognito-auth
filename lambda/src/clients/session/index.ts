/*
 * Copyright (c) 2018 Martin Donath <martin.donath@squidfunk.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

import { CognitoIdentityServiceProvider } from "aws-sdk"

import { Client } from ".."

/* ----------------------------------------------------------------------------
 * Types
 * ------------------------------------------------------------------------- */

/**
 * Session tokens
 */
export interface Session {
  accessToken: string                  /* Access token */
  refreshToken?: string                /* Refresh token */
}

/* ----------------------------------------------------------------------------
 * Class
 * ------------------------------------------------------------------------- */

/**
 * Session client
 */
export class SessionClient extends Client {

  /**
   * Factory method, necessary for mocking
   *
   * @param session - Session
   * @param cognito - Cognito client
   *
   * @return Session client
   */
  public static factory(
    session: Session,
    cognito = new CognitoIdentityServiceProvider({
      apiVersion: "2016-04-18"
    })
  ) {
    return new SessionClient(session, cognito)
  }

  /**
   * Initialize client with session
   *
   * @param session - Session
   * @param cognito - Cognito client
   */
  protected constructor(
    protected session: Session,
    cognito: CognitoIdentityServiceProvider
  ) {
    super(cognito)
  }

  /**
   * Change password for current user
   *
   * @param previous - Previous password
   * @param proposed - Proposed password
   *
   * @return Promise resolving with no result
   */
  public async changePassword(
    previous: string, proposed: string
  ): Promise<void> {
    await this.cognito.changePassword({
      AccessToken: this.session.accessToken,
      PreviousPassword: previous,
      ProposedPassword: proposed
    }).promise()
  }
}
