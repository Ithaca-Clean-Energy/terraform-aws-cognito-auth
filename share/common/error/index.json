{
    "$schema": "http://json-schema.org/draft-06/schema#",
    "additionalProperties": false,
    "description": "Error response",
    "properties": {
        "message": {
            "type": "string"
        },
        "type": {
            "type": "string"
        }
    },
    "required": [
        "message",
        "type"
    ],
    "type": "object"
}

