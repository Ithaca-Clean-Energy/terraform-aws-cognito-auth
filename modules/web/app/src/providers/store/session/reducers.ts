/*
 * Copyright (c) 2018 Martin Donath <martin.donath@squidfunk.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

import { Session } from "common"

import {
  SessionActions,
  SessionActionTypes
} from "./actions"

/* ----------------------------------------------------------------------------
 * Types
 * ------------------------------------------------------------------------- */

/**
 * Session state
 */
export type SessionState = Session<string> | null

/* ----------------------------------------------------------------------------
 * Reducer
 * ------------------------------------------------------------------------- */

/**
 * Session reducer
 *
 * @param state - Session state
 * @param action - Session action
 */
export function session(
  state: SessionState = null,
  action: SessionActions
): SessionState {
  switch (action.type) {
    case SessionActionTypes.INIT:
      const { id, access } = action.session
      return { id, access }
    case SessionActionTypes.TERMINATE:
      return null
    default:
      return state
  }
}
