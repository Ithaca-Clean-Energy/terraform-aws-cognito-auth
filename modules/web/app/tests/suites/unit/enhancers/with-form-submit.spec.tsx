/*
 * Copyright (c) 2018 Martin Donath <martin.donath@squidfunk.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

import { mount } from "enzyme"
import * as React from "react"

import {
  withFormSubmit,
  WithFormSubmit
} from "enhancers"

import { chance, Placeholder } from "_/helpers"
import { mockSession } from "_/mocks/common/session"
import { mockStore } from "_/mocks/providers"
import {
  mockAxiosPostWithError,
  mockAxiosPostWithResult
} from "_/mocks/vendor/axios"

/* ----------------------------------------------------------------------------
 * Tests
 * ------------------------------------------------------------------------- */

/* Form submission enhancer */
describe("enhancers/with-form-submit", () => {

  /* Initialize store */
  const store = mockStore()

  /* Clear store */
  beforeEach(() => {
    store.clearActions()
  })

  /* withFormSubmit */
  describe("withFormSubmit", () => {

    /* { form } */
    describe("{ form }", () => {

      /* Apply enhancer to placeholder component */
      const Component = withFormSubmit()(Placeholder)
      const component = mount(<Component />, {
        context: { store }
      })

      /* Form submission state */
      const form = component
        .find<WithFormSubmit>(Placeholder)
        .prop("form")

      /* Test: should initialize form submission state */
      it("should initialize form submission state", () => {
        expect(form).toEqual({ pending: false, success: false })
      })
    })

    /* { setForm } */
    describe("{ setForm }", () => {

      /* Apply enhancer to placeholder component */
      const Component = withFormSubmit()(Placeholder)
      const component = mount(<Component />, {
        context: { store }
      })

      /* Form submission state reducer */
      const setForm = component
        .find<WithFormSubmit>(Placeholder)
        .prop("setForm")

      /* Test: should update form submission state */
      it("should update form submission state", () => {
        setForm({ pending: true, success: false })
        component.update()
        expect(component.find<WithFormSubmit>(Placeholder).prop("form"))
          .toEqual({ pending: true, success: false })
      })
    })

    /* { submit } */
    describe("{ submit }", () => {

      /* Apply enhancer to placeholder component */
      const Component = withFormSubmit()(Placeholder)
      const component = mount(<Component />, {
        context: { store }
      })

      /* Submit form */
      const submit = component
        .find<WithFormSubmit>(Placeholder)
        .prop("submit")

      /* Test: should prepend API base path to current location */
      it("should prepend API base path to current location", async () => {
        const postMock = mockAxiosPostWithResult()
        await submit()
        expect(postMock).toHaveBeenCalledWith(
          jasmine.stringMatching(process.env.API_BASE_PATH!),
          jasmine.any(Object),
          jasmine.any(Object)
        )
      })

      /* Test: should send request to current location */
      it("should send request to current location", async () => {
        const postMock = mockAxiosPostWithResult()
        await submit()
        expect(postMock).toHaveBeenCalledWith(
          `/${process.env.API_BASE_PATH!}/`,
          jasmine.any(Object),
          jasmine.any(Object)
        )
      })

      /* Test: should send request with provided payload */
      it("should send request with provided payload", async () => {
        const data = { [chance.string()]: chance.string() }
        const postMock = mockAxiosPostWithResult()
        await submit(data)
        expect(postMock).toHaveBeenCalledWith(
          jasmine.any(String),
          data,
          jasmine.any(Object)
        )
      })

      /* Test: should send request with content-type header */
      it("should send request with content-type header", async () => {
        const postMock = mockAxiosPostWithResult()
        await submit()
        expect(postMock).toHaveBeenCalledWith(
          jasmine.any(String),
          jasmine.any(Object),
          jasmine.objectContaining({
            headers: {
              "Content-Type": "application/json"
            }
          })
        )
      })

      /* should send request with credentials */
      it("should send request with credentials", async () => {
        const postMock = mockAxiosPostWithResult()
        await submit()
        expect(postMock).toHaveBeenCalledWith(
          jasmine.any(String),
          jasmine.any(Object),
          jasmine.objectContaining({
            withCredentials: true
          })
        )
      })

      /* Test: should dismiss notification before sending request */
      it("should dismiss notification before sending request", async () => {
        mockAxiosPostWithResult()
        await submit()
        expect(store.getActions()).toMatchSnapshot()
      })

      /* with successful request */
      describe("with successful request", () => {

        /* Test: should set response */
        it("should set response", async () => {
          const data = { [chance.string()]: chance.string() }
          mockAxiosPostWithResult(data)
          await submit()
          component.update()
          expect(component.find<WithFormSubmit>(Placeholder).prop("form"))
            .toEqual(jasmine.objectContaining({
              response: data
            }))
        })

        /* Test: should enable success flag */
        it("should enable success flag", async () => {
          mockAxiosPostWithResult()
          await submit()
          component.update()
          expect(component.find<WithFormSubmit>(Placeholder).prop("form"))
            .toEqual(jasmine.objectContaining({
              pending: false,
              success: true
            }))
        })
      })

      /* with failed request */
      describe("with failed request", () => {

        /* Test: should set error */
        it("should set error", async () => {
          const errMock = new Error()
          mockAxiosPostWithError(errMock)
          await submit()
          component.update()
          expect(component.find<WithFormSubmit>(Placeholder).prop("form"))
            .toEqual(jasmine.objectContaining({
              err: errMock
            }))
        })

        /* Test: should disable success flag */
        it("should disable success flag", async () => {
          mockAxiosPostWithError()
          await submit()
          component.update()
          expect(component.find<WithFormSubmit>(Placeholder).prop("form"))
            .toEqual(jasmine.objectContaining({
              pending: false,
              success: false
            }))
        })

        /* Test: should display notification */
        it("should display notification", async () => {
          const errMock = new Error()
          mockAxiosPostWithError(errMock, "__MESSAGE__")
          await submit()
          expect(store.getActions()).toMatchSnapshot()
        })
      })
    })

    /* with target */
    describe("with target", () => {

      /* { submit } */
      describe("{ submit }", () => {

        /* Target URL */
        const target = chance.string()

        /* Apply enhancer to placeholder component */
        const Component = withFormSubmit({ target })(Placeholder)
        const component = mount(<Component />, {
          context: { store }
        })

        /* Submit form */
        const submit = component
          .find<WithFormSubmit>(Placeholder)
          .prop("submit")

        /* Test: should prepend API base path to provided target URL */
        it("should prepend API base path to provided target URL", async () => {
          const postMock = mockAxiosPostWithResult()
          await submit()
          expect(postMock).toHaveBeenCalledWith(
            jasmine.stringMatching(process.env.API_BASE_PATH!),
            jasmine.any(Object),
            jasmine.any(Object)
          )
        })

        /* Test: should send request to provided target URL */
        it("should send request to provided target URL", async () => {
          const postMock = mockAxiosPostWithResult()
          await submit()
          expect(postMock).toHaveBeenCalledWith(
            `/${process.env.API_BASE_PATH!}/${target}`,
            jasmine.any(Object),
            jasmine.any(Object)
          )
        })
      })
    })

    /* with message */
    describe("with message", () => {

      /* { submit } */
      describe("{ submit }", () => {

        /* Message rendered on success */
        const message = "__MESSAGE__"

        /* Apply enhancer to placeholder component */
        const Component = withFormSubmit({ message })(Placeholder)
        const component = mount(<Component />, {
          context: { store }
        })

        /* Submit form */
        const submit = component
          .find<WithFormSubmit>(Placeholder)
          .prop("submit")

        /* Test: should display notification */
        it("should display notification", async () => {
          mockAxiosPostWithResult()
          await submit()
          expect(store.getActions()).toMatchSnapshot()
        })
      })
    })

    /* with authorization */
    describe("with authorization", () => {

      /* Session */
      const session = mockSession()

      /* Apply enhancer to placeholder component */
      const Component = withFormSubmit({ authorize: true })(Placeholder)
      const component = mount(<Component />, {
        context: {
          store: mockStore({ session })
        }
      })

      /* Submit form */
      const submit = component
        .find<WithFormSubmit>(Placeholder)
        .prop("submit")

      /* Test: should send request with authorization header */
      it("should send request with authorization header", async () => {
        const postMock = mockAxiosPostWithResult()
        await submit()
        expect(postMock).toHaveBeenCalledWith(
          jasmine.any(String),
          jasmine.any(Object),
          jasmine.objectContaining({
            headers: jasmine.objectContaining({
              Authorization: "Bearer __TOKEN__"
            })
          })
        )
      })
    })
  })
})
