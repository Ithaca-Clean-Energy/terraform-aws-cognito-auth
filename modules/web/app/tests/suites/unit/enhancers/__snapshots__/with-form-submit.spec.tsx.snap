// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enhancers/with-form-submit withFormSubmit { submit } should dismiss notification before sending request 1`] = `
Array [
  Object {
    "type": "NOTIFICATION_DISMISS",
  },
]
`;

exports[`enhancers/with-form-submit withFormSubmit { submit } with failed request should display notification 1`] = `
Array [
  Object {
    "type": "NOTIFICATION_DISMISS",
  },
  Object {
    "data": Object {
      "message": "__MESSAGE__",
      "type": "ERROR",
    },
    "type": "NOTIFICATION_DISPLAY",
  },
]
`;

exports[`enhancers/with-form-submit withFormSubmit with message { submit } should display notification 1`] = `
Array [
  Object {
    "type": "NOTIFICATION_DISMISS",
  },
  Object {
    "data": Object {
      "message": "__MESSAGE__",
      "type": "SUCCESS",
    },
    "type": "NOTIFICATION_DISPLAY",
  },
]
`;
